### PR TITLE
Improvements to the documentation with regards to offsets

### PR DESCRIPTION
**Pull request by stefanobaghino-da**
changelog_begin
[Docs] Improvements to the documentation with regards to offsets
changelog_end

- Simplify wording to explain the usage of `blockingForEach` in the Java
bindings quickstart page
- Describe offsets in prose in the page about the Ledger API services
- Suggest how to perform crash recovery with offsets on the application
architecture page
- Link to the relevant Protobuf reference docs where approriate

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: docs/source/app-dev/services.rst [hunk 1]
```diff
@@ -159,6 +159,19 @@ Verbosity
 
 See :ref:`verbosity` above.
 
+.. note::
+
+  The RPCs exposed as part of the transaction and active contracts services make use of offsets.
+
+  An offset is an ever-growing number assigned by the participant to each transaction as they are received from the ledger.
+  This number is opaque to the client and must be used as-is (i.e. cliend-side transformations like incrementing it are not guaranteed to produce a valid offset).
```
Comment (cocreature) [inline docs/source/app-dev/services.rst:167]:
```suggestion
  This number is opaque to the client and must be used as-is (i.e. client-side transformations like incrementing it are not guaranteed to produce a valid offset).
```

Comment (cocreature) [reply docs/source/app-dev/services.rst:167]: I believe we guarantee that they can be lexicographically ordered. Worth mentioning?

File: docs/source/app-dev/services.rst [hunk 2]
```diff
@@ -159,6 +159,19 @@ Verbosity
...
+  The state of a ledger (i.e. the set of active contracts) as exposed by the Ledger API is valid at a specific offset, which is why the last message your application receives when calling the ``ActiveContractsService`` is precisely that offset.
+  In this way, the client can keep track of the relevant state without needing to invoke the ``ActiveContractsService`` again, by starting to read transactions from the given offset.
+
+  Offsets are also useful to perform crash recovery and fail over as documented more in depth in the :ref:`application architecture <dealing-with-failures>` page.
```
Comment (cocreature) [inline docs/source/app-dev/services.rst:171]:
```suggestion
  Offsets are also useful to perform crash recovery and failover as documented more in depth in the :ref:`application architecture <dealing-with-failures>` page.
```
https://en.wikipedia.org/wiki/Failover

**Review by cocreature — APPROVED** [review]

File: docs/source/app-dev/services.rst (hunk 1 above)
Comment (stefanobaghino-da) [reply docs/source/app-dev/services.rst]: Mmmm, we do and we mention it in the (linked) referenced docs where they are just an array of byte. Here I call them just "ever-growing number"s for simplicity. Saying that they can be lexicographically ordered seems redundant. Does that make sense?

Comment (cocreature) [reply docs/source/app-dev/services.rst]: I think ever-growing number is somewhat misleading. You basically give the impression of something with very strong guarantees (e.g., consecutive) and then step back in the next sentence. I personally find it more intuitive to go with "they’re completely opaque, you can only order them lexicographically” but both work so your choice.

Comment (stefanobaghino-da) [reply docs/source/app-dev/services.rst]: Better? https://github.com/digital-asset/daml/pull/10180/commits/e857a7cf3f62c30816b79de67ea0dcb689a125a3

Comment (stefanobaghino-da) [reply docs/source/app-dev/services.rst]: I'll assume yes, let me know if you prefer me to change the wording.

**Pull request merged**